bpf/lib/encap.h

```diff
@@ -295,5 +295,90 @@ encap_and_redirect_netdev(struct __ctx_buff *ctx, struct tunnel_key *k,
 }
 #endif /* TUNNEL_MODE */
 
+#if defined(ENABLE_DSR) && DSR_ENCAP_MODE == DSR_ENCAP_GENEVE
+static __always_inline int
+__encap_with_nodeid_opt6(struct __ctx_buff *ctx, __u32 tunnel_endpoint,
+			 __u32 seclabel, __u32 dstid, __u32 vni,
+			 struct geneve_dsr_opt6 *gopt,
+			 enum trace_reason ct_reason,
+			 __u32 monitor,
+			 int *ifindex)
+{
+	int ret_opt;
+	int ret = __encap_with_nodeid(ctx, tunnel_endpoint, seclabel, dstid,
+				      vni, ct_reason, monitor,
+				      ifindex);
+	if (unlikely(ret < 0))
+		return ret;
+
+	ret_opt = ctx_set_encap_opt6(ctx, gopt);
+	if (unlikely(ret_opt < 0))
+		return ret_opt;
+
+	return ret;
+}
+
+static __always_inline int
+__encap_with_nodeid_opt4(struct __ctx_buff *ctx, __u32 tunnel_endpoint,
+			 __u32 seclabel, __u32 dstid, __u32 vni,
+			 struct geneve_dsr_opt4 *gopt,
+			 enum trace_reason ct_reason,
+			 __u32 monitor,
+			 int *ifindex)
+{
+	int ret_opt;
+	int ret = __encap_with_nodeid(ctx, tunnel_endpoint, seclabel, dstid,
+				      vni, ct_reason, monitor,
+				      ifindex);
+	if (unlikely(ret < 0))
+		return ret;
+
+	ret_opt = ctx_set_encap_opt4(ctx, gopt);
+	if (unlikely(ret_opt < 0))
+		return ret_opt;
+
+	return ret;
+}
```

Comment:
One thing I realized while considering in-XDP support - I believe that in the end, we want `ctx_set_encap_info()` to contain both the `ctx_set_tunnel_key()` and the `ctx_set_tunnel_opt()` operations. For TC it doesn't make any difference to keep them split up. But for XDP we need the total length of the encap headers (including the options), so that we can adjust the headroom in a single operation.

So essentially pipe additional `void *opt, __u32 opt_len` parameters through, and skip the `skb_set_tunnel_opt()` if `opt_len == 0`.

Just noting this here so we don't forget. Not sure if it would simplify your current implementation.

Reply:
Is it something like this? I think it is worth trying to simplify the current implementation

```c
static __always_inline int
__encap_with_nodeid(struct __ctx_buff *ctx, __be32 tunnel_endpoint,
		    __u32 seclabel, __u32 dstid, __u32 vni __maybe_unused,
		    enum trace_reason ct_reason, __u32 monitor, int *ifindex
		    void *opt, __u32 opt_len)
{
	__u32 node_id;
	int ret;

	/* When encapsulating, a packet originating from the local host is
	 * being considered as a packet from a remote node as it is being
	 * received.
	 */
	if (seclabel == HOST_ID)
		seclabel = LOCAL_NODE_ID;

	node_id = bpf_ntohl(tunnel_endpoint);

	cilium_dbg(ctx, DBG_ENCAP, node_id, seclabel);

	ret = ctx_set_encap_info(ctx, node_id, seclabel, dstid, vni, ifindex, opt, opt_len);
	if (ret == CTX_ACT_REDIRECT)
		send_trace_notify(ctx, TRACE_TO_OVERLAY, seclabel, dstid, 0, *ifindex,
				  ct_reason, monitor);

	return ret;
}

static __always_inline __maybe_unused int
ctx_set_encap_info(struct __sk_buff *ctx, __u32 node_id, __u32 seclabel,
		   __u32 dstid __maybe_unused, __u32 vni __maybe_unused,
		   int *ifindex, void *opt, __u32 opt_len)
{
	struct bpf_tunnel_key key = {};
	int ret;

#ifdef ENABLE_VTEP
	if (vni != NOT_VTEP_DST)
		key.tunnel_id = vni;
	else
#endif /* ENABLE_VTEP */
		key.tunnel_id = seclabel;

	key.remote_ipv4 = node_id;
	key.tunnel_ttl = IPDEFTTL;

	ret = ctx_set_tunnel_key(ctx, &key, sizeof(key), BPF_F_ZERO_CSUM_TX);
	if (unlikely(ret < 0))
		return DROP_WRITE_ERROR;

	if (opt_len > 0) {
		 int ret = ctx_set_tunnel_opt(ctx, opt, opt_len);
		if (unlikely(ret < 0))
			return DROP_WRITE_ERROR;
	}

	*ifindex = ENCAP_IFINDEX;

	return CTX_ACT_REDIRECT;
}
```

Reply:
>Is it something like this? I think it is worth trying to simplify the current implementation

yep that matches what I had in mind :+1:.

If it already makes your current implementation simpler, then let's go for it :).